### PR TITLE
[Keyboard] Added BOP support

### PR DESCRIPTION
**Pull request by blewis308**
## Description
Added support for the existing bop keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -76,3 +76,7 @@ __pycache__
 
 # Allow to exist but don't include it in the repo
 user_song_list.h
+BOP/.vs/kb/v16/.suo
+*.db
+BOP/.vs/VSWorkspaceState.json
+BOP/.vscode/settings.json
```
Comment (drashna) [inline .gitignore:82]: Please revert these changes. 
```suggetion
```

File: keyboards/bop/bop.c [hunk 2]
```diff
@@ -0,0 +1 @@
+#include "bop.h"
```
Comment (drashna) [inline keyboards/bop/bop.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/bop/bop.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#ifndef BOP_H
+#define BOP_H
```
Comment (drashna) [inline keyboards/bop/bop.h:2]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

```suggestion
#pragma once
```

File: keyboards/bop/bop.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+#include "quantum.h"
+
+#define KEYMAP( \
```
Comment (drashna) [inline keyboards/bop/bop.h:6]:
```suggestion
#define LAYOUT( \
```

File: keyboards/bop/bop.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+) { \
+	{ K000,  K001,  K002,  K003,  K004,  K005,  K006,  K007,  K008,  K009,  K010,  K011,  K012,  K013,  K014,  K015,  K016,  K017,  K018,  K019 }, \
+	{ K100,  K101,  K102,  K103,  K104,  K105,  K106,  K107,  K108,  K109,  K110,  K111,  K112,  K113,  K114,  K115,  K116,  K117,  K118,  K119 }, \
+	{ K200,  K201,  K202,  K203,  K204,  K205,  K206,  K207,  K208,  K209,  K210,  K211,  K212,  K213,  K214,  K215,  K216,  K217,  K218,  K219 }, \
+	{ K300,  K301,  K302,  K303,  K304,  K305,  K306,  K307,  K308,  K309,  K310,  K311,  K312,  K313,  K314,  K315,  K316,  K317,  K318,  K319 }, \
+	{ K400,  K401,  K402,  K403,  K404,  K405,  K406,  K407,  K408,  K409,  K410,  K411,  K412,  K413,  K414,  K415,  K416,  K417,  K418,  K419 }, \
+	{ K500,  K501,  K502,  K503,  K504,  K505,  K506,  K507,  K508,  K509,  K510,  K511,  K512,  K513,  K514,  K515,  K516,  K517,  K518,  K519 }  \
+}
+
+#endif
```
Comment (drashna) [inline keyboards/bop/bop.h:22]:
```suggestion
```

File: keyboards/bop/config.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#ifndef CONFIG_H
+#define CONFIG_H
```
Comment (drashna) [inline keyboards/bop/config.h:2]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

```suggestion
#pragma once
```

File: keyboards/bop/config.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x666b // fk
+#define PRODUCT_ID      0x626F // bo
+#define DEVICE_VER      0x0001 // rev 1
```
Comment (drashna) [inline keyboards/bop/config.h:9]:
```suggestion
#define VENDOR_ID       0x666B // fk
#define PRODUCT_ID      0x626F // bo
#define DEVICE_VER      0x0001 // rev 1
```

File: keyboards/bop/config.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+#define MATRIX_COL_PINS { D5, C5, B0, B1, B2, B3, B4, B5, B6, E7, E6, F0, F7, F6, F5, F4, F3, F2, F1, C6 }
+//    If your board is spamming the end column, change C7 to C6 in the line above and short those pins on the controller
+                        
+#define UNUSED_PINS
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCe 5
```
Comment (drashna) [inline keyboards/bop/config.h:28]: Case matters for defines. 
```suggestion
#define DEBOUNCE 5
```

File: keyboards/bop/config.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* prevent stuck modifiers */
+#define PREVENT_STUCK_MODIFIERS
+
+#endif
```
Comment (drashna) [inline keyboards/bop/config.h:39]: This setting should be removed, as it is deprecated, as it is the default behavior, now. 
```suggestion
```

File: keyboards/bop/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,41 @@
+#include "bop.h"
```
Comment (drashna) [inline keyboards/bop/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/bop/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	KEYMAP(
```
Comment (drashna) [inline keyboards/bop/keymaps/default/keymap.c:6]:
```suggestion
	[0] = LAYOUT(
```

File: keyboards/bop/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+void matrix_init_user(void) {
+}
+
+void matrix_scan_user(void) {
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+	return true;
+}
```
Comment (drashna) [inline keyboards/bop/keymaps/default/keymap.c:41]:
```suggestion
```

File: keyboards/bop/rules.mk [hunk 13]
```diff
@@ -0,0 +1,7 @@
+MCU = at90usb1286
+
+BOOTMAGIC_ENABLE ?= yes
+MOUSEKEY_ENABLE ?= yes
+EXTRAKEY_ENABLE ?= yes
+CONSOLE_ENABLE ?= yes
+NKRO_ENABLE ?= yes
```
Comment (drashna) [inline keyboards/bop/rules.mk:7]:
```suggestion
MCU = at90usb1286

# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
SWAP_HANDS_ENABLE = no
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/bop/rules.mk [hunk 14]
```diff
@@ -0,0 +1,22 @@
+MCU = at90usb1286
+
+# Bootloader selection
+BOOTLOADER = atmel-dfu
```
Comment (drashna) [inline keyboards/bop/rules.mk:4]: And just to make sure, if this is a teensy++ 2.0, then this is the wrong bootloader.  In that case, you want: 
```suggestion
BOOTLOADER = halfkay
```

Otherwise, atmel-dfu will cover most of the avr dfu type bootloaders.

Comment (blewis308) [reply keyboards/bop/rules.mk:4]: Ah got it, thanks

File: keyboards/bop/bop.h [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+	K400, K401, K402, K403, K404, K405, K406, K407, K408, K409, K410, K411, K412, K413, K414, K415, K416, K417, K418, K419, \
+	K500, K501, K502, K503, K504, K505, K506, K507, K508, K509, K510, K511, K512, K513, K514, K515, K516, K517, K518, K519  \
+) { \
+	{ K000,  K001,  K002,  K003,  K004,  K005,  K006,  K007,  K008,  K009,  K010,  K011,  K012,  K013,  K014,  K015,  K016,  K017,  K018,  K019 }, \
+	{ K100,  K101,  K102,  K103,  K104,  K105,  K106,  K107,  K108,  K109,  K110,  K111,  K112,  K113,  K114,  K115,  K116,  K117,  K118,  K119 }, \
+	{ K200,  K201,  K202,  K203,  K204,  K205,  K206,  K207,  K208,  K209,  K210,  K211,  K212,  K213,  K214,  K215,  K216,  K217,  K218,  K219 }, \
+	{ K300,  K301,  K302,  K303,  K304,  K305,  K306,  K307,  K308,  K309,  K310,  K311,  K312,  K313,  K314,  K315,  K316,  K317,  K318,  K319 }, \
+	{ K400,  K401,  K402,  K403,  K404,  K405,  K406,  K407,  K408,  K409,  K410,  K411,  K412,  K413,  K414,  K415,  K416,  K417,  K418,  K419 }, \
+	{ K500,  K501,  K502,  K503,  K504,  K505,  K506,  K507,  K508,  K509,  K510,  K511,  K512,  K513,  K514,  K515,  K516,  K517,  K518,  K519 }  \
+}
```
Comment (drashna) [inline keyboards/bop/bop.h:35]: It needs this, or else it won't compile: 
```suggestion
}

```

**Review by drashna — APPROVED** [review]

File: .gitignore [hunk 16]
```diff
@@ -75,4 +75,4 @@ __pycache__
 /util/fmpp*
 
 # Allow to exist but don't include it in the repo
-user_song_list.h
+user_song_list.h
```
Comment (drashna) [inline .gitignore:78]: The change to this file should be reverted too. Which is mostly adding a line at the end, it looks like.

File: keyboards/bop/readme.md [hunk 17]
```diff
@@ -0,0 +1,16 @@
+# bop
+
+![Bop](https://i.imgur.com/mEWmMgv.jpg)
+![Bop-top](https://i.imgur.com/DbjVdvV.jpg)
```
Comment (fauxpark) [inline keyboards/bop/readme.md:4]:
```suggestion
![Bop](https://i.imgur.com/mEWmMgvh.jpg)
![Bop-top](https://i.imgur.com/DbjVdvVh.jpg)
```

File: keyboards/bop/rules.mk [hunk 18]
```diff
@@ -0,0 +1,22 @@
+MCU = at90usb1286
```
Comment (fauxpark) [inline keyboards/bop/rules.mk:1]:
```suggestion
# MCU name
MCU = at90usb1286
```

File: keyboards/bop/readme.md [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+
+A huge 6x20 ortholinear keyboard inspired by the BFO-9000
+
+* Keyboard Maintainer: [Fruit](https://github.com/Blewis308)
+* Hardware Supported: Bop v1.1 PCB
```
Comment (fauxpark) [inline keyboards/bop/readme.md:9]:
```suggestion
* Hardware Supported: Bop v1.1 PCB, Teensy 2.0++
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/bop/readme.md (hunk 19 above)
Comment (drashna) [reply keyboards/bop/readme.md]: "Teensy++ 2.0", technically.

File: .gitignore (hunk 16 above)
Comment (drashna) [reply .gitignore:78]:
```suggestion
user_song_list.h

```

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!
